walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -39,30 +39,67 @@
                 .map(e -> e.getWhen())
                 .collect(Collectors.toList());
         Collections.sort(eventTimeRanges, TimeRange.ORDER_BY_START);
-        
-        // Check for invalid input of duration.
-        if(duration > TimeRange.WHOLE_DAY.duration()) {
-            return validTimeRanges;
+
+        // First get the time ranges that work for all attendees. 
+        List<TimeRange> eventTimeRangesWithAllAttendees = 
+            getValidEventTimeRanges(events, request.getAllAttendees());
+        Collection<TimeRange> meetingTimeRangesWithAllAttendees = 
+            getMeetingTimeRanges(eventTimeRangesWithAllAttendees, duration);
+
+        // If there are no time ranges that work for all attendees, get the time
+        // ranges that work for only the mandatory attendees.
+        if (meetingTimeRangesWithAllAttendees.size() == 0) {
+            List<TimeRange> eventTimeRangesWithMandatoryAttendees = 
+                getValidEventTimeRanges(events, request.getAttendees());
+            
+            // In the case that there are optional attendees but no mandatory attendees, 
+            // treat the optional attendees as the mandatory attendees
+            if (eventTimeRangesWithMandatoryAttendees.size() == 0 && eventTimeRangesWithAllAttendees.size() != 0 ){
+                return meetingTimeRangesWithAllAttendees;
+            }
+
+            Collection<TimeRange> meetingTimeRangesWithMandatoryAttendees = 
+                getMeetingTimeRanges(eventTimeRangesWithMandatoryAttendees, duration);
+            return meetingTimeRangesWithMandatoryAttendees;
+        }else{
+            return meetingTimeRangesWithAllAttendees;
         }
-        // If there are no scheduled events, the meeting can be scheduled at any time. 
-        if (eventTimeRanges.size() == 0){
-            validTimeRanges.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY,TimeRange.END_OF_DAY,true));
-        }else {
-            for (TimeRange eventTimeRange: eventTimeRanges) {
-                end = eventTimeRange.start();
-                if (end-start >= duration) {
-                    validTimeRanges.add(TimeRange.fromStartEnd(start,end,false));
-                }
-                start = eventTimeRange.end();
-                if (eventTimeRange.end() > lastEnd) {
-                    lastEnd = eventTimeRange.end();
-                }
+    }
+
+    // Get all the valid event time ranges and sort them by the start time. 
+    // Only add to the list if at least one attendee from the event is 
+    // a required attendee in the meeting request. (Otherwise, the event is irrelevant.)
+    public List<TimeRange> getValidEventTimeRanges(Collection<Event> events, Collection<String> requestAttendees) {
```

Comment:
Consider renaming to "getConflictingEventTimeRanges" or "getBlockedTimeRanges" or something that makes it clear that these are the times you cannot select.

I initially read these as the open time ranges because of the "valid" terminology.

Reply:
Okay!